### PR TITLE
[REVIEW] Add CMake option for per-thread default stream

### PR DESCRIPTION
**Pull request by rongou**
This change adds the option to build cuDF with `--default-stream per-thread` enabled.

For code not built with `nvcc`, we pass `-DCUDA_API_PER_THREAD_DEFAULT_STREAM` to gcc.

This is the alternative solution to https://github.com/rapidsai/rmm/issues/352.

By default the option is disabled. To enable it:
```
cmake .. -DPER_THREAD_DEFAULT_STREAM=ON ...
mvn clean install -DPER_THREAD_DEFAULT_STREAM=ON
```
I've tested this manually on some spark jobs. For TPCH Q4, there is about a 25% speedup in my setup (YMMV).

The corresponding change in RMM is https://github.com/rapidsai/rmm/pull/354.

@harrism @jrhemstad @revans2 @jlowe

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by revans2** [comment]
The code looks fine to enable/disable the feature.  I'm mostly concerned about synchronization with cnmem.

**Comment by rongou** [comment]
CNMeM does extra synchronization in PTDS mode: https://github.com/NVIDIA/cnmem/blob/master/src/cnmem.cpp#L386

So far in my testing I haven't found any problems. Once this is merged we can probably switch it on in our CI and test more thoroughly.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -804,6 +804,13 @@ if(HT_DEFAULT_ALLOCATOR)
     set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -DHT_DEFAULT_ALLOCATOR")
 endif(HT_DEFAULT_ALLOCATOR)
 
+option(PER_THREAD_DEFAULT_STREAM "Build with per-thread default stream" OFF)
```
Comment (harrism) [inline cpp/CMakeLists.txt:807]: I think you'll need this in the tests and benchmarks cmakelists also.

Comment (harrism) [reply cpp/CMakeLists.txt:807]: And you need to add a changelog before CI will run.

**Comment by harrism** [comment]
Add to whitelist

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (rongou) [reply cpp/CMakeLists.txt]: Done.

**Comment by harrism** [comment]
@rongou need a Java codeowners review...

**Comment by rongou** [comment]
@jlowe I couldn't figure out a way to link cmake options, so added a section to the readme. Hopefully this is only temporary, we may want to make PTDS the default.

**Pull request merged**